### PR TITLE
Fix timezone standard is ignored

### LINKED ISSUE #12151: Timezone standard is ignored
### Description

World timezone standards exist, including zones UTC-12 to UTC+14, however PrestaShop chooses to ignore these standards and apply antiquated GMT identifiers and country codes instead. It's unclear when this works in practice, but at least according to **PHP Version 7.1.8** the behaviour is flawed. Trying to install PrestaShop fails with a error message:

> Your default timezone is not supported by PHP. Check for typos in your php.ini file and have a look at the list of deprecated timezones at http://php.net/manual/en/timezones.others.php.

### Reproducing

1. Install PrestaShop on a host with PHP 7.1.8
2. Begin configuration (by navigating to 'install-dev/index.php')
3. Click through 'Choose your language' and 'License agreements'
4. Observe the stated error in 'System compatibility'

### Screenshots

A screenshot exists, but it seems that this GitHub system fails to allow easy upload of images.

## Additional information

PrestaShop version: develop
PHP version: 7.1.8
Platform: Openshift Online

### PR DESCRIPTION
**Pull request by michaesc**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This request solves the problem stated in bug report #12151
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | Might break compatibility with end of life versions of PHP
| Deprecations? | No deprecations
| Fixed ticket? | Fixes #12151
| How to test?  | Follow the reproducible instructions in the bug report.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/12152)

### CODE REVIEW

**Comment by michaesc** [comment]
> Your pull request description seems to be incomplete or malformed:
>
No, but you failed to parse the actual response which states that the problem belongs to more than one category.

> * The `category` should be one of: `FO`, `BO`, `CO`, `IN`, `TE`, `WS`, `LO`
>
Not really, but if you insist on receiving less information (leading to not understanding the full scope of this contribution) then I have reduced the information to conform with the request.

> Would you mind completing the contribution table ? This would help us understand how interesting your contribution is.
>
No, I don't mind. In fact, I already did that in the main pull request message text.

**Comment by rdy4ever** [comment]
> > Your pull request description seems to be incomplete or malformed:
> 
> No, but you failed to parse the actual response which states that the problem belongs to more than one category.
> 
> > * The `category` should be one of: `FO`, `BO`, `CO`, `IN`, `TE`, `WS`, `LO`
> 
> Not really, but if you insist on receiving less information (leading to not understanding the full scope of this contribution) then I have reduced the information to conform with the request.
> 
> > Would you mind completing the contribution table ? This would help us understand how interesting your contribution is.
> 
> No, I don't mind. In fact, I already did that in the main pull request message text.

Are you actually arguing with the prestonBot? 👍

**Comment by michaesc** [comment]
>>> Your pull request description seems to be incomplete or malformed:
>> No, but you failed to parse the actual response which states that the problem belongs to more than one category.
>>> * The `category` should be one of: `FO`, `BO`, `CO`, `IN`, `TE`, `WS`, `LO`
>> 
>> Not really, but if you insist on receiving less information (leading to not understanding the full scope of this contribution) then I have reduced the information to conform with the request.
>>> Would you mind completing the contribution table ? This would help us understand how interesting your contribution is.
>> 
>> No, I don't mind. In fact, I already did that in the main pull request message text.
>
> Are you actually arguing with the prestonBot?
>
The prestonbot brought it on herself. I had to dispute her menacing tone and emphasize the bodacious nature of my report.

**Comment by Quetzacoalt91** [comment]
Well, if we at the way it works, I wonder why we make these checks. As it seems we never take in consideration the current configuration the PHP ini file, we set it instead from the available configuration (fixtures or database).

During CLI installation: https://github.com/PrestaShop/PrestaShop/blob/f7168008b48893423b11d17e8ef41b1d940be79e/install-dev/classes/controllerConsole.php#L115

During Web installation: https://github.com/PrestaShop/PrestaShop/blob/f7168008b48893423b11d17e8ef41b1d940be79e/install-dev/classes/controllerHttp.php#L192-L210

On run: https://github.com/PrestaShop/PrestaShop/blob/f7168008b48893423b11d17e8ef41b1d940be79e/config/config.inc.php#L145

To be confirmed, but it seems we could get rid of this `if()` statement.

**Comment by michaesc** [comment]
> Well, if we at the way it works, I wonder why we make these checks. As it seems we never take in consideration the current configuration the PHP ini file, we set it instead from the available configuration (fixtures or database).
>
> [...]
>
> To be confirmed, but it seems we could get rid of this `if()` statement.
>
The timezone handling in head and all tags I checked is not well done, and my bugfix does not solve the problem at the root. For this reason I labeled the pull request as a work around, however I agree it should be confirmed and at the minimum corrected with the one liner I've supplied.

It is personal taste how that one line (in install-dev/init.php) should look, with a _if statement_ or without. Do we want to protect the user from poor administration practices of PaaS providers? The best solution of all involves thoroughly studying timezone handling in PrestaShop, comparing with what modern PHP installations require, and possibly researching contending timezone logic (in case more than one implementation method exists.)

Let's please merge the one line, and leave the real work to a future effort.

**Comment by Quetzacoalt91** [comment]
With your changes, you cover your specific case with `GMT`, but have you checked what happens if you pick another value from http://php.net/manual/en/timezones.others.php? Exactly the same issue.

Removing the `if()` seems to allow any configuration to pass this check, and does not prevent the timezone to be set properly by the installer. I checked by choosing Poland as country, I got `Europe/Warsaw` when I dumped `date_default_timezone_get()` from PrestaShop.

That's why your PR seems not to be enough to fix the error you're talking about in the linked issue.

**Comment by michaesc** [comment]
> With your changes, you cover your specific case with `GMT`, but have you checked what happens if you pick another value from http://php.net/manual/en/timezones.others.php? Exactly the same issue.
>
> Removing the `if()` seems to allow any configuration to pass this check, and does not prevent the timezone to be set properly by the installer. I checked by choosing Poland as country, I got `Europe/Warsaw` when I dumped `date_default_timezone_get()` from PrestaShop.
>
> That's why your PR seems not to be enough to fix the error you're talking about in the linked issue.
>
This is partially true, as it could be that an administrator chooses any crazy value for the timezone such as 'moontime.' This is the reason I pleaded for a full solution which probably involves validation testing of any provided timezone, according to ISO standards.

However, when installing PrestaShop on a PaaS from the github repository (regardless of head or tag) the timezone defaults to 'GMT' which makes this the most urgent target to find and replace. That is why I hardcoded the value there. Maybe better would be a regex including other typical GMT values.

The real solution involves validation testing according to the international standard, and possibly doing this in any code that touches timezone input.

**Comment by Quetzacoalt91** [comment]
> This is partially true, as it could be that an administrator chooses any crazy value for the timezone such as 'moontime.'

That's the opposite, a crazy value seems to not trigger any warning the installation process. This shows that setting any value does not change anything to the installation, because it is not used.

Your changes is for only one "hardcoded" value from a group of possible options, and the other still trigger the error you reported when defined . As-is, I'm afraid this PR cannot be merged. No need to implement a complex system for a value which seems to be unused, when we can get rid of it in all cases.

**Comment by michaesc** [comment]
>> This is partially true, as it could be that an administrator chooses any crazy value for the timezone such as 'moontime.'
>
> That's the opposite, a crazy value seems to not trigger any warning the installation process. This shows that setting any value does not change anything to the installation, because it is not used.
>
I don't understand what you mean, although I have not tried setting 'moontime' and then running the installer on the platform that fails when 'GMT' is set. It's a silly example of course.

> Your changes is for only one "hardcoded" value from a group of possible options, and the other still trigger the error you reported when defined . As-is, I'm afraid this PR cannot be merged. No need to implement a complex system for a value which seems to be unused, when we can get rid of it in all cases.
>
Are you suggesting that I rework the pull request and simple remove the entire line condition, so that the installer works with UTC time in all cases? If yes, then I'll do it.

**Comment by Quetzacoalt91** [comment]
> I don't understand what you mean, although I have not tried setting 'moontime' and then running the installer on the platform that fails when 'GMT' is set. It's a silly example of course.

Sorry if I wasn't clear. I tried to set `date.timezone=Nonsense` in my php.ini, to see how the installer behaves in that case.
It appears no error is triggered and the whole installation process can run successfully. From my understanding, PrestaShop actually does not care about this value.

> Are you suggesting that I rework the pull request and simple remove the entire line condition, so that the installer works with UTC time in all cases? If yes, then I'll do it.

If you want to keep the contribution, this is what I suggest, yes. In case you do not have the time to do it, I can do it by myself. Not asking you first would be an offense to the time you already spent on that issue. :)

**Comment by michaesc** [comment]
>> I don't understand what you mean, although I have not tried setting 'moontime' and then running the installer on the platform that fails when 'GMT' is set. It's a silly example of course.
>
> Sorry if I wasn't clear. I tried to set `date.timezone=Nonsense` in my php.ini, to see how the installer behaves in that case.
> It appears no error is triggered and the whole installation process can run successfully. From my understanding, PrestaShop actually does not care about this value.
>
In addition to 'Nonsense', if you try setting 'GMT' and still get no error then your environment does not allow for reproduction of this bug. It's possible that you got a false positive result. Can you try 'GMT' and report again please?

>> Are you suggesting that I rework the pull request and simple remove the entire line condition, so that the installer works with UTC time in all cases? If yes, then I'll do it.
>
> If you want to keep the contribution, this is what I suggest, yes. In case you do not have the time to do it, I can do it by myself. Not asking you first would be an offense to the time you already spent on that issue. :)
>
Yes, that was very nice of you. Obviously I want to be more involved in the project if possible, and this is a good way to do that.

I've adjusted my pull request code according to your suggestion, which indeed solves more of the problem through the workaround method we agree on.

**Comment by mickaelandrieu** [comment]
> No, I don't mind. In fact, I already did that in the main pull request message text.

And @prestonBot (my virtual child) thank you for that.

Troll apart, can we improve the message generated by Preston? We don't want to hurt anyone here, but we have some rules that help us speed our reviews. As you can see, we have a lot of pull requests to review and asking some information to the contributors help us to figure out the interest of a pull request.

Mickaël

**Review by Quetzacoalt91 — DISMISSED** [review]

**Comment by Quetzacoalt91** [comment]
> In addition to 'Nonsense', if you try setting 'GMT' and still get no error then your environment does not allow for reproduction of this bug. It's possible that you got a false positive result. Can you try 'GMT' and report again please?

* Tried 'GMT' without your changes: The error appears and the installation is blocked.
* Tried 'GMT' with your changes: All green!

Tested on PHP 5.6 & 7.2. It seems there is no false positive on my side. :)

**Comment by michaesc** [comment]
>> In addition to 'Nonsense', if you try setting 'GMT' and still get no error then your environment does not allow for reproduction of this bug. It's possible that you got a false positive result. Can you try 'GMT' and report again please?
>
> * Tried 'GMT' without your changes: The error appears and the installation is blocked.
> * Tried 'GMT' with your changes: All green!
>
> Tested on PHP 5.6 & 7.2. It seems there is no false positive on my side. :)
>
Good news that you can reproduce the error, but bad news that it has been present since PrestaShop on PHP 5.6.

Thank you for accepting my change request, and I'll be very happy once it passes QA and is merged into trunk.

File: install-dev/init.php [hunk 1]
```diff
@@ -123,10 +123,10 @@
...
 
 @set_time_limit(0);
-if (!@ini_get('date.timezone')) {
-    @date_default_timezone_set('Europe/Paris');
-    ini_set('date.timezone', 'UTC');
-}
+// Work around lack of validation for timezone
+// standards conformance, mandatory in PHP 7
+@date_default_timezone_set('Europe/Paris');
+ini_set('date.timezone', 'UTC');
```
Comment (PierreRambaud) [inline install-dev/init.php]: We could not force everyone to be in UTC.
Maybe try something like
```php
if (!in_array(@ini_get('date.timezone'), timezone_identifiers_list()) {
    @date_default_timezone_set('Europe/Paris');
    ini_set('date.timezone', 'UTC');
}
```

**Review by PierreRambaud — DISMISSED** [review]
Don't force UTC for everyone

File: install-dev/init.php (hunk 1 above)
Comment (eternoendless) [reply install-dev/init.php]: Watch out, these are two different timezones.
```php
@date_default_timezone_set('Europe/Paris');
ini_set('date.timezone', 'UTC');
```

To be coherent, it should be:
```php
@date_default_timezone_set('UTC');
ini_set('date.timezone', 'UTC');
```

Comment (michaesc) [reply install-dev/init.php]: @eternoendless and @PierreRambaud, your conclusion is correct and your suggestion is good. I'm reengineering the original hack, and testing the next set of changes before modifying the pull request.

**Review by PierreRambaud — APPROVED** [review]

File: install-dev/init.php (hunk 1 above)
Comment (michaesc) [reply install-dev/init.php]: @PierreRambaud, your suggestion contained a fatal syntax error which I've corrected. The new code integrates @eternoendless's suggestion as well, and corrects all problems known in this pull request thread to date.

I think the pull request can now be merged, and request a code review to confirm this.

**Comment by mbadrani** [comment]
I think we need more explanations to reproducing this behavior because me @ntiepresta and @khouloudbelguith failed to reproducing this, (trying with "Nonsense" and "GMT" values on the php.ini file)

**Comment by michaesc** [comment]
> I think we need more explanations to reproducing this behavior because me @ntiepresta and @khouloudbelguith failed to reproducing this, (trying with "Nonsense" and "GMT" values on the php.ini file)
>
There are many ways to choose a PHP hosting platform where the problem is impossible to reproduce, for example those that forbit administrative access and force a PHP environment (containing a timezone.)

The way I found the bug was by creating a Docker cointainer and running it on OpenShift. If you want to try this, there is a good tutorial (mostly still works after OpenShift updated from version 2 to 3) right in the forum:

https://www.prestashop.com/forums/topic/394953-how-to-install-prestashop-on-openshift-by-red-hat/

If others explain how they reproduced the bug it would be useful to document that here, so please everybody be forthcoming with your information.

**Comment by eternoendless** [comment]
Defaults change to UTC should have to be done here as well, either in this PR or in a follow-up.

https://github.com/PrestaShop/PrestaShop/blob/65dfa13e3876d711ed8e2779fedef89cdbaa08dc/install-dev/upgrade/upgrade.php#L79-L81

Also in the autoupgrade module:

https://github.com/PrestaShop/autoupgrade/blob/0e679ea94e460b4970094ebf37c33524533f3cee/classes/UpgradeTools/CoreUpgrader/CoreUpgrader.php#L162-L164

**Comment by michaesc** [comment]
> Defaults change to UTC should have to be done here as well, either in this PR or in a follow-up.
>
Let's correct the problems in install-dev/upgrade/upgrade.php in a separate pull request.

The autoupgrade correction belongs in a completely different repository (called github.com/PrestaShop/autoupgrade/.) I think I'll leave that up to somebody else, because my impression is that there is a lot more to correct than a couple hardcoded timezones.

It might be helpful to create a bug report in that repository.

**Comment by PierreRambaud** [comment]
Thanks @michaesc

**Pull request merged**